language/stdlib/modules/account_limits.move

```diff
@@ -0,0 +1,240 @@
+    // `limits_definition.time_period` has elapsed, resets the window and
+    // the inflow and outflow records. Additionally the new
+    // `tracked_balance` is computed at this time as well.
+    fun reset_window(window: &mut Window, limits_definition: &LimitsDefinition) {
+        Transaction::assert(limits_definition.is_certified, 1);
+        let current_time = current_time();
+        if (current_time > window.window_start + limits_definition.time_period) {
+            window.window_start = current_time;
+            window.window_inflow = 0;
+            window.window_outflow = 0;
+        }
+    }
+
+    // Verify that the receiving account tracked by the `receiving` window
+    // can receive `amount` funds without violating requirements
+    // specified the `limits_definition` passed in.
+    fun can_receive<CoinType>(
+        amount: u64,
+        receiving: &mut Window,
+        limits_definition: &LimitsDefinition
+    ): bool {
+        // If the limits ares unrestricted then no more work needs to be done
+        if (is_unrestricted(limits_definition)) return true;
+        reset_window(receiving, limits_definition);
+        // Check that the max inflow is OK
+        let inflow_ok = receiving.window_inflow + amount <= limits_definition.max_inflow;
+        // Check that the holding after the deposit is OK
+        let holding_ok = receiving.tracked_balance + amount <= limits_definition.max_holding;
+        // The account with `receiving` window can receive the payment so record it.
+        if (inflow_ok && holding_ok) {
+            receiving.window_inflow = receiving.window_inflow + amount;
+            receiving.tracked_balance = receiving.tracked_balance + amount;
+        };
+        inflow_ok && holding_ok
+    }
+
+    // Verify that `amount` can be withdrawn from the account tracked
+    // by the `sending` window without violating any limits specified by
+    // the passed-in `limits_definition`.
+    fun can_withdraw<CoinType>(
+        amount: u64,
+        sending: &mut Window,
+        limits_definition: &LimitsDefinition
+    ): bool {
+        // If the limits are unrestricted then no more work is required
+        if (is_unrestricted(limits_definition)) return true;
+        reset_window(sending, limits_definition);
+        // Check max outlflow
+        let outflow = sending.window_outflow + amount;
+        let outflow_ok = outflow <= limits_definition.max_outflow;
+        // Outflow is OK, so record it.
+        if (outflow_ok) {
+            sending.window_outflow = outflow;
+            sending.tracked_balance = if (amount >= sending.tracked_balance) 0
+                                       else sending.tracked_balance - amount;
+        };
+        outflow_ok
+    }
+
+    // Return whether the LimitsDefinition definition is unrestricted or
+    // not.
+    fun is_unrestricted(limits_def: &LimitsDefinition): bool {
+        limits_def.max_inflow == 0 &&
```

Comment:
Would all U64_MAX instead of all zero make this look like less of a special case? The short-circuiting based on `is_unrestricted` seems valuable, but would be nice if the code still behaved correctly if we forgot to short-circuit.

Reply:
I went back and forth on this...I think the point you make is a good case for why it should be max instead of min. I'll update.